rust/src/delta.rs

```diff
@@ -613,6 +662,143 @@ impl std::fmt::Debug for DeltaTable {
+
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+    options: DeltaTransactionOptions,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    pub fn new(delta_table: &'a mut DeltaTable, options: Option<DeltaTransactionOptions>) -> Self {
+        DeltaTransaction { 
+            delta_table,
+            options: options.map_or_else(|| DeltaTransactionOptions { 
+                max_retry_commit_attempts: DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS 
+            }, |o| o),
+        }
+    }
+
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // TODO: stubbing `operation` parameter (which will be necessary for writing the CommitInfo action), but leaving it unused for now.
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // Serialize all actions that are part of this log entry.
+        let mut jsons = Vec::<String>::new();
+
+        for action in actions {
+            let json = serde_json::to_string(action)?;
+            jsons.push(json);
+        }
+        
+        let log_entry = jsons.join("\n");
+        let log_entry = log_entry.as_bytes();
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(log_entry).await?;
+
+        // NOTE: since we have the log entry in memory already, 
+        // we could optimize this further by merging the log entry instead of updating from storage.
+        self.delta_table.update().await?;
+
+        Ok(version)
+    }
+
+    async fn try_commit_loop(&mut self, log_entry: &[u8]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            let commit_result = self.try_commit(log_entry).await;
+
+            if commit_result.is_ok() {
+                return commit_result;
+            }
+
+            attempt_number += 1;
+
+            if attempt_number > self.options.max_retry_commit_attempts + 1 {
+                return Err(DeltaTransactionError::CommitRetriesExceeded);
```

Comment:
Related to the above comment, perhaps in the case of retries being exhausted, the `CommitRetriesExceeded` could carry an `inner` Error out from `try_commit`?

Basically I'm thinking of the remote storage backends where commit errors are likely to be much more common, and how the caller could be left in the dark without propagating this information out.

Reply:
@rtyler, regarding this comment and the one above, I have [decomposed the error types and added an inner to CommitRetriesExceeded](https://github.com/delta-io/delta-rs/pull/84/files#diff-3b5c80af9e5527fbeaa6dad151a835f0733a26e61d0ba0471c58f243a29dfa99R666).

I'll be happy to add error/warn logging as well, but will wait for @houqp to chime in since you addressed him directly.